File: android/src/main/kotlin/dev/note11/flutter_naver_map/flutter_naver_map/model/overlay/overlay/NPolylineOverlay.kt
package dev.note11.flutter_naver_map.flutter_naver_map.model.overlay.overlay

import androidx.annotation.ColorInt
import com.naver.maps.geometry.LatLng
import com.naver.maps.map.overlay.Overlay
import com.naver.maps.map.overlay.PolylineOverlay
import com.naver.maps.map.overlay.PolylineOverlay.LineCap
import com.naver.maps.map.overlay.PolylineOverlay.LineJoin
import dev.note11.flutter_naver_map.flutter_naver_map.converter.AddableOverlay
import dev.note11.flutter_naver_map.flutter_naver_map.converter.DefaultTypeConverter.asDouble
import dev.note11.flutter_naver_map.flutter_naver_map.converter.DefaultTypeConverter.asInt
import dev.note11.flutter_naver_map.flutter_naver_map.converter.DefaultTypeConverter.asList
import dev.note11.flutter_naver_map.flutter_naver_map.converter.DefaultTypeConverter.asMap
import dev.note11.flutter_naver_map.flutter_naver_map.converter.MapTypeConverter.asLatLng
import dev.note11.flutter_naver_map.flutter_naver_map.converter.MapTypeConverter.asLineCap
import dev.note11.flutter_naver_map.flutter_naver_map.converter.MapTypeConverter.asLineJoin
import dev.note11.flutter_naver_map.flutter_naver_map.converter.MapTypeConverter.toMap
import dev.note11.flutter_naver_map.flutter_naver_map.converter.MapTypeConverter.toMessageableString
import dev.note11.flutter_naver_map.flutter_naver_map.model.enum.NOverlayType
import dev.note11.flutter_naver_map.flutter_naver_map.model.overlay.NOverlayInfo
import dev.note11.flutter_naver_map.flutter_naver_map.util.DisplayUtil
import kotlin.math.roundToInt

internal data class NPolylineOverlay(
    override val info: NOverlayInfo,
    val coords: List<LatLng>,
    @ColorInt val color: Int,
    val widthDp: Double,
    val lineCap: LineCap,
    val lineJoin: LineJoin,
    val patternDp: List<Int>,
) : AddableOverlay<PolylineOverlay> {

    override fun createMapOverlay(): PolylineOverlay = PolylineOverlay().also { g ->
        g.coords = coords
        g.color = color
        g.width = DisplayUtil.dpToPx(widthDp)
        g.capType = lineCap
        g.joinType = lineJoin
        g.setPattern(*patternDp.map { DisplayUtil.dpToPx(it.toDouble()) }.toIntArray())
    }

    override fun toMap(): Map<String, Any?> = mapOf(
        infoName to info.toMap(),
        coordsName to coords.map { it.toMap() },
        colorName to color,
        widthName to widthDp,
        lineCapName to lineCap.toMessageableString(),
        lineJoinName to lineJoin.toMessageableString(),
        patternName to patternDp.map { DisplayUtil.pxToDp(it).roundToInt() },
    )

    companion object {
        fun fromMap(rawMap: Any): NPolylineOverlay = rawMap.asMap().let {
            NPolylineOverlay(
                info = NOverlayInfo.fromMap(it[infoName]!!),
                coords = it[coordsName]!!.asList { latLng -> latLng.asLatLng() },
                color = it[colorName]!!.asInt(),
                widthDp = it[widthName]!!.asDouble(),
                lineCap = it[lineCapName]!!.asLineCap(),
                lineJoin = it[lineJoinName]!!.asLineJoin(),
                patternDp = it[patternName]!!.asList { p -> p.asInt() },
            )
        }

        fun fromPolylineOverlay(
            polylineOverlay: Overlay,
            id: String,
        ): NPolylineOverlay = (polylineOverlay as PolylineOverlay).run {
            NPolylineOverlay(
                info = NOverlayInfo(NOverlayType.POLYLINE_OVERLAY, id),
                coords = coords,
                color = color,
                widthDp = DisplayUtil.pxToDp(width),
                lineCap = capType,
                lineJoin = joinType,
                patternDp = pattern.map { DisplayUtil.pxToDp(it).roundToInt() }
            )
        }

        /*
            --- Messaging Name Define ---
        */

        private const val infoName = "info";
        const val coordsName = "coords";
        const val colorName = "color";
        const val widthName = "width";
        const val lineCapName = "lineCap";
        const val lineJoinName = "lineJoin";
        const val patternName = "pattern";
        const val boundsName = "bounds";
    }
}

File: ios/Classes/model/map/overlay/overlay/NPolylineOverlay.swift
import NMapsMap

internal struct NPolylineOverlay: AddableOverlay {
    typealias OverlayType = NMFPolylineOverlay
    var overlayPayload: Dictionary<String, Any?> = [:]

    let info: NOverlayInfo
    let coords: Array<NMGLatLng>
    let minZoom: Double?
    let maxZoom: Double?
    let color: UIColor
    let width: Double
    let lineCap: NMFOverlayLineCap
    let lineJoin: NMFOverlayLineJoin
    let pattern: Array<NSNumber>

    func createMapOverlay() -> OverlayType {
        let polyline = NMFPolylineOverlay()
        polyline.line = NMGLineString(points: coords)
        if let minZoom = minZoom {
            polyline.minZoom = minZoom
        }
        if let maxZoom = maxZoom {
            polyline.maxZoom = maxZoom
        }
        polyline.color = color
        polyline.width = CGFloat(width)
        polyline.capType = lineCap
        polyline.joinType = lineJoin
        polyline.pattern = pattern
        return polyline
    }

    static func fromMessageable(_ v: Any) -> NPolylineOverlay {
        let d = asDict(v)
        return NPolylineOverlay(
                info: NOverlayInfo.fromMessageable(d[infoName]!),
                coords: asArr(d[coordsName]!, elementCaster: asLatLng),
                minZoom: castOrNull(d[minZoomName], caster: asDouble),
                maxZoom: castOrNull(d[maxZoomName], caster: asDouble),
                color: asUIColor(d[colorName]!),
                width: asDouble(d[widthName]!),
                lineCap: asLineCap(d[lineCapName]!),
                lineJoin: asLineJoin(d[lineJoinName]!),
                pattern: asArr(d[patternName]!) {
                    NSNumber(value: asRoundInt(rawFloat: $0))
                }
        )
    }

    /*
        --- Messaging Name Define ---
    */
    private static let infoName = "info"
    static let coordsName = "coords";
    static let colorName = "color";
    static let widthName = "width";
    static let lineCapName = "lineCap";
    static let lineJoinName = "lineJoin";
    static let patternName = "pattern";
    static let boundsName = "bounds";
}
